File: src/phase2/linkedlist/doublylinkedlist/LruCache.java
package phase2.linkedlist.doublylinkedlist;

import phase2.linkedlist.doublylinkedlist.Node;

import java.util.HashMap;
import java.util.Map;

import static phase2.linkedlist.Node.printLinkedList;
import static phase2.linkedlist.doublylinkedlist.Node.printDoublyLinkedList;

public class LruCache {
    /**
     * 1. Need a set to verify hit or miss
     * 2. Need a DoublyLinkedList for managing sequence
     */

    private Map<Integer, Node> cache = new HashMap<>();
    private Node head;
    private Node tail;

    private int count;

    private int capacity;

    public LruCache(int initialValue, int capacity) {
        Node node = new Node(initialValue);
        cache.put(initialValue, node);
        head = node;
        tail = node;
        this.capacity = capacity;
    }

    public void add(int value) {
        //check if present in map
        if (cache.containsKey(value)) {
            //make this node as the head if it is not the head
            Node node = cache.get(value);
            moveTheNodeToHeadPosition(node);
        } else {
            //put in the map and make this node as the head if he
            //and remove last node if capacity is reached.
            //increase count
            Node node = new Node(value);
            cache.put(value, node);
            insertNodeAtHeadPosition(node);
            count+=1;
            if (count >= capacity) {
                tail = tail.prev;
                tail.next = null;
                count-=1;
            }
        }
        System.out.println();
        printDoublyLinkedList(head);
        System.out.println();
    }

    private void moveTheNodeToHeadPosition(Node node) {
        if (tail == node && tail.prev != null) {
            tail = tail.prev;
            tail.next = null;
            node.prev = null;
            node.next = head;
            head.prev = node;
            head = node;
            return;
        }
        if (head != node) {
            node.prev.next = node.next;
            node.next.prev = node.prev;
            node.prev = null;
            node.next = head;
            head.prev = node;
            head = node;
        }
    }

    private void insertNodeAtHeadPosition(Node node) {
           head.prev = node;
           node.next = head;
           head = node;
           node.prev = null;
    }
}

File: src/phase2/linkedlist/grind75/LRUCache.kt
package phase2.linkedlist.grind75

import kotlin.math.sign

class LRUCache(capacity: Int) {

    private var cache: MutableMap<Int, LruCacheNode> = mutableMapOf()
    private  var head: LruCacheNode? = null
    private  var tail: LruCacheNode? = null
    private var count = 0
    private var capacity = capacity

    fun get(key: Int): Int {
       if (cache.containsKey(key)) {
           val node = cache[key] as LruCacheNode

           if (head == tail || node == head) {
               return node.keyValue.second
           }

           if (node == tail) {
               tail = tail?.prev
               tail?.next = null
               node.next = head
               head?.prev = node
               head = node
               head?.prev = null

           } else {
               node.prev?.next = node.next
               node.next?.prev = node.prev

               node.next = head
               head?.prev = node
               head = node
               node.prev = null
           }
           return node.keyValue.second
       }
        return -1
    }

    fun put(key: Int, value: Int) {
         //check in cache, if present then move it to the head position
         //if not present then add it to the head position and adjust LinkedList as per capacity
         if (cache.containsKey(key)) {
             val node = cache[key] as LruCacheNode
             node.keyValue = Pair(key, value)
             if (head == tail || node == head) {
                 return
             }
             if (node == tail) {
                 tail = tail?.prev
                 tail?.next = null
                 node.next = head
                 head?.prev = node
                 head = node
                 head?.prev = null
             } else {
                 node.prev?.next = node.next
                 node.next?.prev = node.prev

                 node.next = head
                 head?.prev = node
                 head = node
             }
         } else {
             count+=1
             val node = LruCacheNode(Pair(key, value))
             cache[key] = node
             if (head == null) {
                 head = node
                 tail = node
             } else {
                 head?.prev = node
                 node.next = head
                 head = node
                 if (count > capacity) {
                     cache.remove(tail?.keyValue?.first)
                     tail = tail?.prev
                     tail?.next = null
                 }
             }
         }

    }

    class LruCacheNode(var keyValue: Pair<Int, Int>) {

        var prev: LruCacheNode? = null
        var next: LruCacheNode? = null

        init {
            prev = null
            next = null
        }
    }
}
